### PR TITLE
Added go server

### PR DESCRIPTION
**Pull request by mibrah42**
## What are we adding in this PR

This PR adds the initial setup for our go server. 

The go server will be in charge of handling initial rendering of the application and any further interaction with the database. These interactions will include fetching and saving templates to our datastore.

New folder structure:
```
.
├── app.yaml
├── client
├── data
├── go.mod
├── go.sum
├── handlers
├── main.go
└── static
```

The entry point of our application is the main.go file. In the main function, we do the following: 

1. Create a datastore client. 
```go
	/** 
	* Creating a datastore client. This instance is shared across the application.
	*/
	db, err := datastore.NewClient(ctx, os.Getenv("GOOGLE_CLOUD_PROJECT"))
	if err != nil {
		log.Fatal("Database connection error", err)
	}
```

2. Initialize a new ServeMux and register handlers. 

> ServeMux is an HTTP request multiplexer. It matches the URL of each incoming request against a list of registered patterns and calls the handler for the pattern that most closely matches the URL. 
https://golang.org/pkg/net/http/#ServeMux

- Register a static file server on the index route ("/") and map any requests with a "/static/" prefix to our static directory. 
- Register the templates handler to the route `api/v1/templates`.
  - `GET`: Fetches all templates from the database.
  - `POST`: Adds a new template to the database (implemented in a following PR).
- Wrap server in a CORS middleware. Currently accepts all origins but will be fine tuned to only accept app-creator domain.
- Running server in the background.

```go 
	/**
	* Initializing a new server mux to handle routes.
	*/
	serverMux := http.NewServeMux()
	
	/**
	* Serving static files.
	*/
	fs := http.FileServer(http.Dir("./static"))
	serverMux.Handle("/", fs)
	serverMux.Handle("/static/", http.StripPrefix("/static/", http.FileServer(http.Dir("./static"))))
	
	/**
	* Setting up handlers.
	*/
	templatesHandler := handlers.NewTemplatesHandler(l, db)

	/**
	* Registering handlers.
	*/
	serverMux.Handle("/api/v1/templates", templatesHandler)

	/**
	* CORS middleware. Default option allows all origins. 
	* https://github.com/rs/cors
	*/ 
	c := cors.Default().Handler(serverMux)

	/**
	* Setting up server instance.
	*/
	server := &http.Server{
		Addr:         port,
		Handler:      c,
		IdleTimeout:  120 * time.Second,
		ReadTimeout:  1 * time.Second,
		WriteTimeout: 1 * time.Second,
	}

	/**
	* Start server on a new thread. This is a blocking operation
	* so we need to run it in the background.
	*/
	go func() {
		err := server.ListenAndServe()
		if err != nil {
			l.Fatal(err)
		}
	}()
```

3. Handle server shutdown.

When the server is shutdown, we first stop accepting additional requests. If there are requests that are currently in progress, we give them time to finish by setting a timeout of 30 seconds. This ensures that sudden server shutdowns does not affect requests that are currently in progress.

```go
	/**
	* Handle termination gracefully by listening to Interrupt and Kill signals
	* and waiting 30 seconds to make sure any handlers being executed get to
	* complete in time.
	*/
	sigChan := make(chan os.Signal)
	signal.Notify(sigChan, os.Interrupt)
	signal.Notify(sigChan, os.Kill)

	sig := <-sigChan
	l.Println("Shutting down gracefully", sig)

	timeoutContext, cancel := context.WithTimeout(ctx, 30*time.Second)
	defer cancel()

	server.Shutdown(timeoutContext)
```

The next two files to look at are `/handlers/templates/templates.go` and `/data/templates/templates.go`. 

Lets first look at the templates handler. 

In order to create a handler that responds to http requests, all we need is a struct that implements the handler interface.
```go
type Handler interface {
    ServeHTTP(ResponseWriter, *Request)
}
```
https://golang.org/pkg/net/http/#Handler

Here we have the definition of the templates handler struct. We initialize it with a logger and database client pointers and we implement the Handler interface by adding a `ServeHTTP` method to it. 

The `ServeHTTP` method first checks if the request verb is either a `GET` or a `POST`. It then calls  `t.getTemplates(rw, r)` or `t.addTemplate(rw, r)` respectively.

```go
type Templates struct {
	l *log.Logger 
	db *datastore.Client
}

func (t *Templates) ServeHTTP(rw http.ResponseWriter, r *http.Request) {
	if r.Method == http.MethodGet {
		t.getTemplates(rw, r)
		return
	}

	if r.Method == http.MethodPost {
		t.addTemplate(rw, r)
		return
	}

	// Return method-not-allowed status for non-implemented methods.
	rw.WriteHeader(http.StatusMethodNotAllowed)
}
```

Implementation for `getTemplates`, and `addTemplate`
```go
/*
* Helper method called on GET requests to "/templates".
*/
func (t *Templates) getTemplates(rw http.ResponseWriter, r *http.Request) {
	templates := data.GetTemplates(t.db, t.l, r.Context())
	err := templates.ToJSON(rw)
	if err != nil {
		http.Error(rw, "Conversion to JSON was unsuccessful", http.StatusInternalServerError)
	}
}

/*
* Helper method called on POST requests to "/templates". Adds a new template to the database.
*/
func (t *Templates) addTemplate(rw http.ResponseWriter, r *http.Request) {
	template := &data.Template{}
	err := template.FromJSON(r.Body)
	if err != nil {
		http.Error(rw, "Failed JSON conversion", http.StatusBadRequest)
	}

	t.l.Printf("Template: %#v", template)
	data.AddTemplate(template)
}
```

We also have a helper method that returns an instance of the templates handler that we call from `main.go`.
```go
/**
* Creates new templates handler with a provided logger instance.
*/
func NewTemplatesHandler(l *log.Logger, db *datastore.Client) *Templates {
	return &Templates{l, db}
}
```

To abstract the logic of handling requests and data fetching. We have another templates.go file that deals with accessing the datastore. 

Here we have a struct that defines what a template looks like
```go
type Template struct {
	Id string `json:"id"` 
	Name string `json:"name"`
	ImageUrl string `json:"imageUrl"`
	Template string `json:"template"`
}
```

And additional methods for converting a template to and from JSON format. 
```go
/**
* Converts an array of Template data into JSON.
*/
func (t *Templates) ToJSON(w io.Writer) error {
	encoder := json.NewEncoder(w)
	return encoder.Encode(t)
}

/**
* Converts a JSON representation of a Template object into a Template instance. 
*/
func (t *Template) FromJSON(r io.Reader) error {
	e := json.NewDecoder(r)
	return e.Decode(t)
}
```

We also implement a GetTemplates method that accesses the datastore to retrieve templates.
```go
/**
* Returns array of templates.
*/
func GetTemplates(db *datastore.Client, l *log.Logger, ctx context.Context) Templates {
	templates := []*Template{}
	_, err := db.GetAll(ctx, datastore.NewQuery("Template"), &templates)
	if err != nil {
		l.Fatal("Unable to fetch templates\n")
	}

	return templates
}
```

### CODE REVIEW

File: app-creator/main.go [hunk 1]
```diff
@@ -0,0 +1,30 @@
+package main
+
+import (
+	"log"
+	"os"
+
```
Comment (sufyanAbbasi) [inline app-creator/main.go]: Intentional extra space?

File: app-creator/main.go [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+	// Setting static directory.
+	e.Static("/static", "static")
+
+	// Serving index.html on "/" route.
+	e.File("/", "static/index.html")
+
+	// Check if a PORT env variable is set.
+	port := os.Getenv("PORT")
+	if port == "" {
```
Comment (sufyanAbbasi) [inline app-creator/main.go]: Add a regex check for the port value to be all numbers and always remove leading colons if they exist. https://golang.org/pkg/regexp/

**Review by sufyanAbbasi — APPROVED** [review]

File: app-creator/main.go (hunk 1 above)
Comment (mibrah42) [reply app-creator/main.go]: Handled by the editor, separates stdlib imports from external ones.

File: app-creator/main.go (hunk 2 above)
Comment (sufyanAbbasi) [reply app-creator/main.go]: Also it would be good to add tests for this too.

File: app-creator/data/templates/templates.go [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+/**
+* Returns array of templates.
+*/
+func GetTemplates(db *datastore.Client, l *log.Logger, ctx context.Context) Templates {
+	templates := []*Template{}
+	_, err := db.GetAll(ctx, datastore.NewQuery("Template"), &templates)
+	if err != nil {
+		l.Fatal("Unable to fetch templates\n")
+	}
+
```
Comment (sufyanAbbasi) [inline app-creator/data/templates/templates.go]: We may want to support a convention to reorder the templates. This can happen client or server side.

File: app-creator/data/templates/templates.go [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+	e := json.NewDecoder(r)
+	return e.Decode(t)
+}
+
+/**
+* Returns array of templates.
+*/
+func GetTemplates(db *datastore.Client, l *log.Logger, ctx context.Context) Templates {
+	templates := []*Template{}
+	_, err := db.GetAll(ctx, datastore.NewQuery("Template"), &templates)
```
Comment (sufyanAbbasi) [inline app-creator/data/templates/templates.go]: We may want to support a one-time operation to load up the datastore with the default templates if none exist, since users who build the app creator locally will have fresh datastores.

File: app-creator/data/templates/templates_test.go [hunk 5]
```diff
@@ -0,0 +1,149 @@
...
+import (
+	"testing"
+	"net/http/httptest"
+	"io/ioutil"
+	"net/http"
+	"encoding/json"
+	"strings"
+)
+
+func TestToJSON(t *testing.T) {
```
Comment (sufyanAbbasi) [inline app-creator/data/templates/templates_test.go]: I would make this and test name below more descriptive: 

`TestTemplateSerializationToJSON`

`TestTemplateDeserializationFromJSON`

File: app-creator/go.sum [hunk 6]
```diff
@@ -0,0 +1,351 @@
+cloud.google.com/go v0.26.0/go.mod h1:aQUYkXzVsufM+DwF1aE+0xfcU+56JwCaLick0ClmMTw=
```
Comment (sufyanAbbasi) [inline app-creator/go.sum:1]: Is this file necessary to check into the repo? Is it generated automatically?

File: app-creator/handlers/templates/templates_test.go [hunk 7]
```diff
@@ -0,0 +1,42 @@
...
+	req, err := http.NewRequest("GET", "localhost:8080/api/v1/templates", nil)
+	if err != nil {
+		t.Fatalf("could not create request: %v", err)
+	}
+
+	rec := httptest.NewRecorder()
+
+	l := log.New(os.Stdout, "app-creator", log.LstdFlags)
+
+	db, err := datastore.NewClient(req.Context(), os.Getenv("GOOGLE_CLOUD_PROJECT"))
```
Comment (sufyanAbbasi) [inline app-creator/handlers/templates/templates_test.go]: You don't want to rely on the actual datastore client in the test, instead, you should use a stub.
Create an interface for the `db` client that includes the `GetAll` method, then inject a custom object that implements `GetAll` returning fake JSON templates. 

This way, you can also test whether the system properly handles errors by adding a test where `GetAll` throws an error instead, and check that the http server returns a 500 error.

File: app-creator/main.go [hunk 8]
```diff
@@ -0,0 +1,119 @@
...
+	* Strip any leading colons and make sure it only contains numbers.
+	*/
+	re := regexp.MustCompile(`^[:]([0-9]+)$`)
+	
+	/**
+	* Default to port 8080 if environment variable is not set or is invalid.
+	*/
+	if port == "" || !re.MatchString(port) {
+		port = ":8080"
+		log.Printf("Defaulting to port %s", port)
```
Comment (sufyanAbbasi) [inline app-creator/main.go]: You may want to specify a better error log message if the formatting of the port variable is wrong.

File: app-creator/main.go [hunk 9]
```diff
@@ -0,0 +1,119 @@
...
+	*/
+	templatesHandler := handlers.NewTemplatesHandler(l, db)
+
+	/**
+	* Registering handlers.
+	*/
+	serverMux.Handle("/api/v1/templates", templatesHandler)
+
+	/**
+	* CORS middleware. Default option allows all origins.
```
Comment (sufyanAbbasi) [inline app-creator/main.go]: Add an environment variable that, if present, only allows CORS for those domains (in a comma separated list). That way, in the App Engine yaml, we can add the appspot URL and limit the  XSS attack vector.

File: app-creator/main.go [hunk 10]
```diff
@@ -0,0 +1,119 @@
...
+	/**
+	* Handle termination gracefully by listening to Interrupt and Kill signals
+	* and waiting 30 seconds to make sure any handlers being executed get to
+	* complete in time.
+	*/
+	sigChan := make(chan os.Signal)
+	signal.Notify(sigChan, os.Interrupt)
+	signal.Notify(sigChan, os.Kill)
+
+	sig := <-sigChan
```
Comment (sufyanAbbasi) [inline app-creator/main.go:136]: Go is awesome!

**Review by sufyanAbbasi — CHANGES_REQUESTED** [review]

File: app-creator/main.go [hunk 11]
```diff
@@ -0,0 +1,119 @@
...
+
+	/**
+	* Check if a PORT env variable is set.
+	*/
+	port := os.Getenv("PORT")
+
+	/**
+	* Strip any leading colons and make sure it only contains numbers.
+	*/
+	re := regexp.MustCompile(`^[:]([0-9]+)$`)
```
Comment (bengalin) [inline app-creator/main.go]: Why the square brackets in `[:]`?

File: app-creator/handlers/templates/templates.go [hunk 12]
```diff
@@ -0,0 +1,63 @@
...
+*/
+func NewTemplatesHandler(l *log.Logger, db *datastore.Client) *Templates {
+	return &Templates{l, db}
+}
+
+/*
+* Handler called on requests made to "/templates".
+*/
+func (t *Templates) ServeHTTP(rw http.ResponseWriter, r *http.Request) {
+	if r.Method == http.MethodGet {
```
Comment (bengalin) [inline app-creator/handlers/templates/templates.go]: Optional: `switch r.Method { case http.MethodGet: [...] }`.

File: app-creator/handlers/templates/templates.go [hunk 13]
```diff
@@ -0,0 +1,63 @@
...
+
+import (
+	"log"
+	"net/http"
+	data "modules/data/templates"
+	"cloud.google.com/go/datastore"
+)
+
+type Templates struct {
+	l *log.Logger
```
Comment (bengalin) [inline app-creator/handlers/templates/templates.go]: A thought: you can embed the log.Logger; see example and reasoning in https://golang.org/doc/effective_go.html#embedding (`type Job struct`).

File: app-creator/handlers/templates/templates.go [hunk 14]
```diff
@@ -0,0 +1,63 @@
+package handlers
+
+import (
+	"log"
+	"net/http"
+	data "modules/data/templates"
```
Comment (bengalin) [inline app-creator/handlers/templates/templates.go]: Avoid renaming import if possible (https://github.com/golang/go/wiki/CodeReviewComments#imports).

File: app-creator/data/templates/templates.go [hunk 15]
```diff
@@ -0,0 +1,50 @@
+package templates_data
```
Comment (bengalin) [inline app-creator/data/templates/templates.go]: No underscores in package names (https://golang.org/doc/effective_go.html#package-names).  Ideally, this should be the same as the file name, i.e., `templates`.

File: app-creator/data/templates/templates.go [hunk 16]
```diff
@@ -0,0 +1,50 @@
...
+
+import (
+	"encoding/json"
+	"io"
+	"context"
+	"cloud.google.com/go/datastore"
+	"log"
+)
+
+type Template struct {
```
Comment (bengalin) [inline app-creator/data/templates/templates.go]: Please document all exported names (https://github.com/golang/go/wiki/CodeReviewComments#doc-comments)

File: app-creator/data/templates/templates.go [hunk 17]
```diff
@@ -0,0 +1,50 @@
...
+	Id string `json:"id"` 
+	Name string `json:"name"`
+	ImageUrl string `json:"imageUrl"`
```
Comment (bengalin) [inline app-creator/data/templates/templates.go]: `ImageURL` (https://github.com/golang/go/wiki/CodeReviewComments#initialisms)

File: app-creator/data/templates/templates.go [hunk 18]
```diff
@@ -0,0 +1,50 @@
...
+	Template string `json:"template"`
+}
+
+type Templates []*Template
+
+/**
+* Converts an array of Template data into JSON.
```
Comment (bengalin) [inline app-creator/data/templates/templates.go]: Doc comments should begin with the documented name (https://golang.org/doc/effective_go.html#commentary)

File: app-creator/data/templates/templates.go [hunk 19]
```diff
@@ -0,0 +1,50 @@
...
+*/
+func (t *Template) FromJSON(r io.Reader) error {
+	e := json.NewDecoder(r)
+	return e.Decode(t)
+}
+
+/**
+* Returns array of templates.
+*/
+func GetTemplates(db *datastore.Client, l *log.Logger, ctx context.Context) Templates {
```
Comment (bengalin) [inline app-creator/data/templates/templates.go]: Functions/methods that take a context param should have it as their first param (https://github.com/golang/go/wiki/CodeReviewComments#contexts).

Also, for a caller, this would be `templates.GetTemplates` which stutters IMO.  I'd rename to `Get`.

File: app-creator/data/templates/templates_test.go [hunk 20]
```diff
@@ -0,0 +1,149 @@
...
+
+	err := templates.ToJSON(rec)
+	if err != nil {
+		t.Fatalf("error converting to json: %v", err)
+	}
+
+	res := rec.Result()
+	defer res.Body.Close()
+	if res.StatusCode != http.StatusOK {
+		t.Fatalf("expected status OK; got %v", res.StatusCode)
```
Comment (bengalin) [inline app-creator/data/templates/templates_test.go]: Got before want in error messages (https://github.com/golang/go/wiki/CodeReviewComments#useful-test-failures)

File: app-creator/handlers/templates/templates.go [hunk 21]
```diff
@@ -0,0 +1,63 @@
+package handlers
```
Comment (bengalin) [inline app-creator/handlers/templates/templates.go]: I'd rename the file to `handlers.go`.  But this might be different than the naming scheme that you have elsewhere in the project (is it `parent/child/child.ext`?)  Are we going to have other handlers and/or other data types?  If not, then another option is to place all of these files in the same folder (`app-creator/{main|handlers|templates}.go`).

The file name would typically dictate the package name, which would in turn influence the choice for the `Templates` struct name and its `NewTemplatesHandler` constructor below.  My preference would be to make it so that the call site is either `handlers.New()` or `templates.NewHandler()`.

**Review by bengalin — COMMENTED** [review]
Mostly style-related comments.  I added a comment only for the first occurrences, so please address throughout.

File: app-creator/go.sum (hunk 6 above)
Comment (mibrah42) [reply app-creator/go.sum:1]: > Ensure your go.sum file is committed along with your go.mod file. See FAQ below for more details and rationale.
https://github.com/golang/go/wiki/Modules#releasing-modules-all-versions

The checksum helps ensure that whoever is running the project can have the exact version of the dependencies.

File: app-creator/handlers/templates/templates.go (hunk 21 above)
Comment (mibrah42) [reply app-creator/handlers/templates/templates.go:1]: For now, we only have the templates handler. I thought having a directory for handlers would be the best approach if we want to add more handlers in the future. I have updated the directory to look like the following.

```bash
├──data
│   ├── data.go
│   └── data_test.go
├── handlers
│   ├── handlers.go
│   └── handlers_test.go
├── main.go
```

Comment (mibrah42) [reply app-creator/handlers/templates/templates.go]: For now, we only have the templates handler. I thought having a directory for handlers would be the best approach if we want to add more handlers in the future. So something like this: 

```
├── handlers
│   ├── <handler_one>.go
│   └── <handler_one>_test.go
│   ├── <handler_two>.go
│   └── <handler_two>_test.go
```

 I have updated the directory to look like the following.

```bash
├──data
│   ├── data.go
│   └── data_test.go
├── handlers
│   ├── handlers.go
│   └── handlers_test.go
├── main.go
```

File: app-creator/handlers/templates/templates.go (hunk 13 above)
Comment (mibrah42) [reply app-creator/handlers/templates/templates.go]: I'm using the logger in another file data.go. If I want to embed the logger, I will have to move this logic to the templates.go file along with other values. Otherwise, I will have circular imports. Will it be better to move it over?
```
func Get(ctx context.Context, db *datastore.Client, l *log.Logger) Templates {
	...
	if err != nil {
		l.Fatal("Unable to fetch templates\n")
	}
        ...
}
```

File: app-creator/data/templates/templates.go (hunk 3 above)
Comment (mibrah42) [reply app-creator/data/templates/templates.go]: I think having it on the client will be the better approach. I will handle it in the fetching-templates PR.

File: app-creator/data/templates/templates.go (hunk 4 above)
Comment (mibrah42) [reply app-creator/data/templates/templates.go]: Since we are making changes to the template structure, I will add this in a following PR.

Comment (mibrah42) [reply app-creator/data/templates/templates.go]: issue tracked here https://github.com/googleinterns/earthengine-community/projects/1#card-41793493

File: app-creator/main.go [hunk 22]
```diff
@@ -0,0 +1,145 @@
...
+
+	/*
+	  Check if a PORT env variable is set.
+	*/
+	port := os.Getenv("PORT")
+
+	/*
+	  Strip any leading colons and make sure it only contains numbers.
+	*/
+	re := regexp.MustCompile(`^:([0-9]+)$`)
```
Comment (sufyanAbbasi) [inline app-creator/main.go:51]: Make the leading colon optional: `^:?([0-9]+)$`

File: app-creator/main.go [hunk 23]
```diff
@@ -0,0 +1,145 @@
...
+	*/ 
+	c := cors.Default().Handler(serverMux)
+
+	var allowedOrigins []string
+	allowedOriginsEnv := os.Getenv("ALLOWED_ORIGINS")
+	// If env variable is set for allowedOrigins, use them instead of the default.
+	if allowedOriginsEnv != "" {
+		// Split comma separated domains and trim whitespaces.
+		allowedOrigins = strings.Split(allowedOriginsEnv, ",")
+		for i := 0; i < len(allowedOrigins); i++ {
```
Comment (sufyanAbbasi) [inline app-creator/main.go]: There's a slightly nicer way of performing a map in Go: https://gobyexample.com/collection-functions
```
for i, o := range allowedOrigins {
  allowedOrigins[i] = strings.Trim(o, " ")
}
```

File: app-creator/data/data.go [hunk 24]
```diff
@@ -0,0 +1,57 @@
...
+}
+
+/**
+ Get is used to fetch templates from the database and returns a slice of templates.
+*/
+func Get(ctx context.Context, db dsiface.Client, l *log.Logger) Templates {
+	templates := []*Template{}
+	_, err := db.GetAll(ctx, datastore.NewQuery("Template"), &templates)
+	if err != nil {
+		l.Fatal("Unable to fetch templates\n")
```
Comment (bengalin) [inline app-creator/data/data.go]: Rather than logging, WDYT about returning the error to the caller and have it decide what to do?

Basically, I am proposing this signature:
`func Get(ctx context.Context, db dsiface.Client) (Templates, error)`

Typically, libraries do not log errors, but rather bubble them up with additional context, if helpful.

File: app-creator/main.go (hunk 22 above)
Comment (sufyanAbbasi) [inline app-creator/main.go]: Make the leading colon optional: `^:?([0-9]+)$`

**Review by sufyanAbbasi — APPROVED** [review]
Just a few more small changes. I would recommend reaching out to @bengalin to see if he can do a quick once over.

File: app-creator/data/data_test.go [hunk 25]
```diff
@@ -0,0 +1,149 @@
...
+	"encoding/json"
+	"strings"
+)
+
+func TestToJSON(t *testing.T) {
+	rec := httptest.NewRecorder()
+
+	err := TemplatesStub.ToJSON(rec)
+	if err != nil {
+		t.Fatalf("error converting to json: %v", err)
```
Comment (bengalin) [inline app-creator/data/data_test.go]: nit: Please use consistent sentence capitalization in log and error messages here and below.  The convention is to capitalize the initial in `log.*` and `t.*`, but not to capitalize `fmt.Error` and `errors.New`.  See https://github.com/golang/go/wiki/CodeReviewComments#error-strings for explanation.

File: app-creator/data/data.go [hunk 26]
```diff
@@ -0,0 +1,57 @@
...
+	}
+
+	return templates
+}
+
+// TODO: Implement adding templates to the database. Used when a user saves a template.
+func AddTemplate(t *Template) {}
```
Comment (bengalin) [inline app-creator/data/data.go]: This is another potentially error-returning function, IMO.

File: app-creator/handlers/handlers.go [hunk 27]
```diff
@@ -0,0 +1,67 @@
...
+ Adds a new template to the database.
+*/
+func (t *Templates) addTemplate(rw http.ResponseWriter, r *http.Request) {
+	template := &data.Template{}
+	err := template.FromJSON(r.Body)
+	if err != nil {
+		http.Error(rw, "Failed JSON conversion", http.StatusBadRequest)
+	}
+
+	t.l.Printf("Template: %#v", template)
```
Comment (bengalin) [inline app-creator/handlers/handlers.go]: nit: Maybe `Added template: %#v`?  Also, is there any other context that would be useful to log?  I'd also log only after `AddTemplate` returned successfully (and log an error if it didn't).

File: app-creator/handlers/handlers_test.go [hunk 28]
```diff
@@ -0,0 +1,45 @@
...
+	dbclient := dsiface.AdaptClient(db);
+
+	th := NewHandler(l, dbclient)
+
+	th.ServeHTTP(rec, req)
+
+	res := rec.Result()
+	defer res.Body.Close()
+	if res.StatusCode != http.StatusOK {
+		t.Errorf("expected status OK; got %v", res.StatusCode)
```
Comment (bengalin) [inline app-creator/handlers/handlers_test.go:38]: Got before want.

File: app-creator/handlers/handlers.go [hunk 29]
```diff
@@ -0,0 +1,67 @@
...
+*/
+type Templates struct {
+	l *log.Logger 
+	db dsiface.Client
+}
+
+/*
+ NewHandler creates a new templates handler with a provided logger instance.
+*/
+func NewHandler(l *log.Logger, db dsiface.Client) *Templates {
```
Comment (bengalin) [inline app-creator/handlers/handlers.go]: I'd rename to just `New` so that the caller does `handlers.New(..)`.

File: app-creator/handlers/handlers.go [hunk 30]
```diff
@@ -0,0 +1,67 @@
...
+	"log"
+	"net/http"
+	"modules/data"
+	"github.com/googleapis/google-cloud-go-testing/datastore/dsiface"
+)
+
+/*
+ Templates define the structure of a templates handler instance.
+*/
+type Templates struct {
```
Comment (bengalin) [inline app-creator/handlers/handlers.go]: `TemplatesHandler` might be a more appropriate name, albeit wordy.  WDYT?

File: app-creator/main.go [hunk 31]
```diff
@@ -0,0 +1,145 @@
...
+	  Getting application context.
+	*/
+	ctx := context.Background()
+
+	/* 
+	  Creating a datastore client. This instance is shared across the application.
+	*/
+	db, err := datastore.NewClient(ctx, os.Getenv("GOOGLE_CLOUD_PROJECT"))
+	if err != nil {
+		log.Fatal("Database connection error", err)
```
Comment (bengalin) [inline app-creator/main.go]: `log.Fatalf("Database connection error: %v", err)`?

File: app-creator/main.go [hunk 32]
```diff
@@ -0,0 +1,145 @@
...
+	serverMux.Handle("/api/v1/templates", handler)
+
+	/*
+	  CORS middleware. Default option allows all origins. 
+	  https://github.com/rs/cors
+	  ch - cors handler
+	*/ 
+	c := cors.Default().Handler(serverMux)
+
+	var allowedOrigins []string
```
Comment (bengalin) [inline app-creator/main.go]: I don't think we need this variable at this scope, so please declare when it is initialized.

If you want, the block below can be simplified and de-indented by directly instantiating: `allowedOrigins := strings.Split(os.Getenv("ALLOWED_ORIGINS", ","))` and avoiding `allowedOriginsEnv` altogether. Though, pretty sure you'd still want to check if `len(allowedOrigins) != 0` before creating the CORS handler (not sure what an empty `AllowedOrigins` in `cors.Options` would do).

File: app-creator/handlers/handlers.go (hunk 30 above)
Comment (mibrah42) [reply app-creator/handlers/handlers.go]: I agree 👍

File: app-creator/handlers/handlers.go [hunk 33]
```diff
@@ -8,24 +8,24 @@ import (
...
+ TemplatesHandler define the structure of a templates handler instance.
 */
-type Templates struct {
+type TemplatesHandler struct {
 	l *log.Logger 
 	db dsiface.Client
 }
 
 /*
  NewHandler creates a new templates handler with a provided logger instance.
```
Comment (bengalin) [inline app-creator/handlers/handlers.go]: s/NewHandler/New/

**Review by bengalin — APPROVED** [review]
